<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns='http://www.w3.org/1999/xhtml' xml:lang='en' lang='en'>
<head>
<title>ignatenkobrain/dnf-obsoletes/dnf.git - ignatenkobrain/dnf-obsoletes/dnf</title>
<meta name='generator' content='cgit v0.12'/>
<meta name='robots' content='index, nofollow'/>
<link rel='stylesheet' type='text/css' href='/cgit-data/cgit.css'/>
<link rel='shortcut icon' href='/favicon.ico'/>
<link rel='alternate' title='Atom feed' href='http://copr-dist-git.fedorainfracloud.org/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/atom/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch?h=master' type='application/atom+xml'/>
<link rel='vcs-git' href='http://copr-dist-git.fedorainfracloud.org/git/ignatenkobrain/dnf-obsoletes/dnf.git' title='ignatenkobrain/dnf-obsoletes/dnf.git Git repository'/>
</head>
<body>
<div id='cgit'><table id='header'>
<tr>
<td class='logo' rowspan='2'><a href='/cgit/'><img src='/cgit-data/cgit.png' alt='cgit logo'/></a></td>
<td class='main'><a href='/cgit/'>index</a> : <a title='ignatenkobrain/dnf-obsoletes/dnf.git' href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/'>ignatenkobrain/dnf-obsoletes/dnf.git</a></td><td class='form'><form method='get' action=''>
<input type='hidden' name='id' value='2e95f5547fb7ed083fd63a7923922bb587d5f374'/><select name='h' onchange='this.form.submit();'>
<option value='f25'>f25</option>
<option value='master' selected='selected'>master</option>
</select> <input type='submit' name='' value='switch'/></form></td></tr>
<tr><td class='sub'>ignatenkobrain/dnf-obsoletes/dnf</td><td class='sub right'></td></tr></table>
<table class='tabs'><tr><td>
<a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/'>summary</a><a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/refs/?id=2e95f5547fb7ed083fd63a7923922bb587d5f374'>refs</a><a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/log/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch'>log</a><a class='active' href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/tree/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch?id=2e95f5547fb7ed083fd63a7923922bb587d5f374'>tree</a><a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/commit/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch?id=2e95f5547fb7ed083fd63a7923922bb587d5f374'>commit</a><a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/diff/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch?id=2e95f5547fb7ed083fd63a7923922bb587d5f374'>diff</a><a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/stats/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch'>stats</a></td><td class='form'><form class='right' method='get' action='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/log/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch'>
<input type='hidden' name='id' value='2e95f5547fb7ed083fd63a7923922bb587d5f374'/><select name='qt'>
<option value='grep'>log msg</option>
<option value='author'>author</option>
<option value='committer'>committer</option>
<option value='range'>range</option>
</select>
<input class='txt' type='text' size='10' name='q' value=''/>
<input type='submit' value='search'/>
</form>
</td></tr></table>
<div class='path'>path: <a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/tree/?id=2e95f5547fb7ed083fd63a7923922bb587d5f374'>root</a>/<a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/tree/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch?id=2e95f5547fb7ed083fd63a7923922bb587d5f374'>0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch</a></div><div class='content'>blob: aef492b1afcfd88ff77a9e3551d05ebc969ebbfa (<a href='/cgit/ignatenkobrain/dnf-obsoletes/dnf.git/plain/0001-subject-prefer-obsoletes-RhBug-1096506-RhBug-1332830.patch?id=2e95f5547fb7ed083fd63a7923922bb587d5f374'>plain</a>)
<table summary='blob content' class='blob'>
<tr><td class='linenumbers'><pre><a id='n1' href='#n1'>1</a>
<a id='n2' href='#n2'>2</a>
<a id='n3' href='#n3'>3</a>
<a id='n4' href='#n4'>4</a>
<a id='n5' href='#n5'>5</a>
<a id='n6' href='#n6'>6</a>
<a id='n7' href='#n7'>7</a>
<a id='n8' href='#n8'>8</a>
<a id='n9' href='#n9'>9</a>
<a id='n10' href='#n10'>10</a>
<a id='n11' href='#n11'>11</a>
<a id='n12' href='#n12'>12</a>
<a id='n13' href='#n13'>13</a>
<a id='n14' href='#n14'>14</a>
<a id='n15' href='#n15'>15</a>
<a id='n16' href='#n16'>16</a>
<a id='n17' href='#n17'>17</a>
<a id='n18' href='#n18'>18</a>
<a id='n19' href='#n19'>19</a>
<a id='n20' href='#n20'>20</a>
<a id='n21' href='#n21'>21</a>
<a id='n22' href='#n22'>22</a>
<a id='n23' href='#n23'>23</a>
<a id='n24' href='#n24'>24</a>
<a id='n25' href='#n25'>25</a>
<a id='n26' href='#n26'>26</a>
<a id='n27' href='#n27'>27</a>
<a id='n28' href='#n28'>28</a>
<a id='n29' href='#n29'>29</a>
<a id='n30' href='#n30'>30</a>
<a id='n31' href='#n31'>31</a>
<a id='n32' href='#n32'>32</a>
<a id='n33' href='#n33'>33</a>
<a id='n34' href='#n34'>34</a>
<a id='n35' href='#n35'>35</a>
<a id='n36' href='#n36'>36</a>
<a id='n37' href='#n37'>37</a>
<a id='n38' href='#n38'>38</a>
<a id='n39' href='#n39'>39</a>
<a id='n40' href='#n40'>40</a>
<a id='n41' href='#n41'>41</a>
<a id='n42' href='#n42'>42</a>
<a id='n43' href='#n43'>43</a>
<a id='n44' href='#n44'>44</a>
<a id='n45' href='#n45'>45</a>
<a id='n46' href='#n46'>46</a>
</pre></td>
<td class='lines'><pre><code>From 4838eeb486df7bce075da97ea1233935b6688a2e Mon Sep 17 00:00:00 2001
From: Jan Silhan &lt;jsilhan@redhat.com&gt;
Date: Mon, 13 Jun 2016 23:05:44 +0200
Subject: [PATCH] subject: prefer obsoletes (RhBug:1096506)(RhBug:1332830)

DNF makes better selection of packages considering obsoletes from
'pkg_spec' input when `multilib_policy` is set to `best`
(by Subject.set_best_selector()). Moreover the selection of packages
is evaluated during resolution so when the obsoleting package has
broken dependencies the next best choose is picked (RhBug:1096506)

Subject.get_best_selector searches by provides first (RhBug:1332830)
while prioritizing obsoletes and then packages with the same name as a
provide.

Closes: #609
Approved by: j-mracek
---
 dnf/subject.py | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/dnf/subject.py b/dnf/subject.py
index ea7a0e57..a1b3fbf4 100644
--- a/dnf/subject.py
+++ b/dnf/subject.py
@@ -123,9 +123,14 @@ class Subject(object):
         nevra = first(self.subj.nevra_possibilities_real(sack, **kwargs))
         if nevra:
             sltr = dnf.selector.Selector(sack)
-            s = self._nevra_to_selector(sltr, nevra)
-            if len(s.matches()) &gt; 0:
-                return s
+            if nevra._has_just_name():
+                s = sltr.set(provides=nevra.name)
+                if len(s.matches()) &gt; 0:
+                    return s
+            else:
+                s = self._nevra_to_selector(sltr, nevra)
+                if len(s.matches()) &gt; 0:
+                    return s
 
         reldep = first(self.subj.reldep_possibilities_real(sack))
         if reldep:
-- 
2.11.0

</code></pre></td></tr></table>
</div> <!-- class=content -->
<div class='footer'>generated by <a href='http://git.zx2c4.com/cgit/about/'>cgit v0.12</a> at 2017-01-17 15:39:02 (GMT)</div>
</div> <!-- id=cgit -->
</body>
</html>
